File: gcp-gce-ops/assets/code-snippets/create_instance.go
// create_instance.go — Create a GCE VM instance via Go SDK
// Usage: go run ./create_instance.go
package main

import (
    "context"
    "fmt"
    "log"
    "os"
    
    compute "cloud.google.com/go/compute/apiv1"
    "cloud.google.com/go/compute/apiv1/computepb"
    "google.golang.org/api/option"
)

func main() {
    ctx := context.Background()
    project := os.Getenv("CLOUDSDK_CORE_PROJECT")
    zone := os.Getenv("CLOUDSDK_COMPUTE_ZONE")
    if zone == "" { zone = "us-central1-a" }
    instanceName := os.Getenv("GCE_INSTANCE_NAME")
    if instanceName == "" { instanceName = "my-instance" }
    
    client, err := compute.NewInstancesRESTClient(ctx,
        option.WithCredentialsFile(os.Getenv("GOOGLE_APPLICATION_CREDENTIALS")))
    if err != nil {
        log.Fatalf("NewInstancesRESTClient: %v", err)
    }
    defer client.Close()
    
    req := &computepb.InsertInstanceRequest{
        Project: project,
        Zone:    zone,
        InstanceResource: &computepb.Instance{
            Name: protoPtr(instanceName),
            MachineType: protoPtr(fmt.Sprintf("projects/%s/zones/%s/machineTypes/n2-standard-2", project, zone)),
            Disks: []*computepb.AttachedDisk{{
                Boot: protoBool(true),
                InitializeParams: &computepb.AttachedDiskInitializeParams{
                    SourceImage: protoPtr("projects/debian-cloud/global/images/family/debian-12"),
                    DiskSizeGb:  protoInt64(20),
                    DiskType:    protoPtr(fmt.Sprintf("projects/%s/zones/%s/diskTypes/pd-balanced", project, zone)),
                },
            }},
            NetworkInterfaces: []*computepb.NetworkInterface{{
                Name: protoPtr("default"),
            }},
        },
    }
    
    op, err := client.Insert(ctx, req)
    if err != nil { log.Fatalf("Insert: %v", err) }
    if err := op.Wait(ctx); err != nil { log.Fatalf("Wait: %v", err) }
    fmt.Println("Created instance:", instanceName)
}

func protoPtr(s string) *string { return &s }
func protoBool(b bool) *bool { return &b }
func protoInt64(i int64) *int64 { return &i }

File: gcp-gce-ops/assets/code-snippets/create_instance.py
#!/usr/bin/env python3
# create_instance.py — Create a GCE VM instance via Python SDK
# Usage: python3 create_instance.py

import os

from google.cloud import compute_v1

project = os.environ["CLOUDSDK_CORE_PROJECT"]
zone = os.environ.get("CLOUDSDK_COMPUTE_ZONE", "us-central1-a")
instance_name = os.environ.get("GCE_INSTANCE_NAME", "my-instance")

client = compute_v1.InstancesClient()

# Configure boot disk
boot_disk = compute_v1.AttachedDisk()
boot_disk.boot = True
disk_config = compute_v1.AttachedDiskInitializeParams()
disk_config.source_image = "projects/debian-cloud/global/images/family/debian-12"
disk_config.disk_size_gb = 20
disk_config.disk_type = f"projects/{project}/zones/{zone}/diskTypes/pd-balanced"
boot_disk.initialize_params = disk_config

# Configure network
network_interface = compute_v1.NetworkInterface()
network_interface.name = "default"

# Configure instance
instance = compute_v1.Instance()
instance.name = instance_name
instance.machine_type = f"projects/{project}/zones/{zone}/machineTypes/n2-standard-2"
instance.disks = [boot_disk]
instance.network_interfaces = [network_interface]

request = compute_v1.InsertInstanceRequest(
    project=project, zone=zone, instance_resource=instance)
operation = client.insert(request=request)
operation.result(timeout=300)
print(f"Created instance: {instance_name}")
